shared/middleware/files_api.rb

```diff
@@ -174,12 +161,50 @@ def record_event(quota_event_type, quota_type, encrypted_channel_id)
     quota_crossed_half_used(endpoint, encrypted_channel_id) if quota_crossed_half_used?(app_size, body.length)
     response = buckets.create_or_replace(encrypted_channel_id, filename, body, request.GET['version'])
 
-    content_type :json
     category = mime_type.split('/').first
     {filename: filename, category: category, size: body.length, versionId: response.version_id}.to_json
   end
 
   #
+  # PUT /v3/(assets|sources)/<channel-id>/<filename>?version=<version-id>
+  #
+  # Create or replace a file. Optionally overwrite a specific version.
+  #
+  put %r{/v3/sources/([^/]+)/([^/]+)$} do |encrypted_channel_id, filename|
+    dont_cache
+    content_type :json
+
+    # read the entire request before considering rejecting it, otherwise varnish
+    # may return a 503 instead of whatever status code we specify. Unfortunately
+    # this prevents us from rejecting large files based on the Content-Length
+    # header.
+    body = request.body.read
+
+    put_file('sources', encrypted_channel_id, filename, body)
+  end
+
+  # POST /v3/assets/<channel-id>/<filename>?version=<version-id>
+  #
+  # Upload a new file. We use this method so that IE9 can still upload by
+  # posting to an iframe.
+  #
+  post %r{/v3/assets/([^/]+)/new$} do |encrypted_channel_id|
```

Comment:
/new didn't feel right to me, but I couldnt quite put my finger on why. I think I was thinking that posting to /assets/channel_id/ should create a new channel, but of course that's a completely separate API. will update.